Use `libqrencode` for MSA QR codes

getchoo opened the pull request:
Follow up on https://github.com/PrismLauncher/PrismLauncher/pull/3257

`qrcodegen` isn't available in most repositories, package discovery
scripts for it are maintained in a different repository, and they are
bugged for Windows at least. This basically forces us into vendoring it,
which isn't cool; libqrencode seems like viable alternative used by many
more apps

Signed-off-by: Seth Flynn <getchoo@tuta.io>

blocked on #3977

getchoo committed:
feat(ui/MSALoginDialog): use libqrencode for qr codes

`qrcodegen` isn't available in most repositories, package discovery
scripts for it are maintained in a different repository, and they are
bugged for Windows at least. This basically forces us into vendoring it,
which isn't cool; libqrencode seems like viable alternative used by many
more apps

Signed-off-by: Seth Flynn <getchoo@tuta.io>

getchoo commented:
![image](https://github.com/user-attachments/assets/d3e25eaa-5609-49c0-a737-efa3e06e45e5)

Trial97 commented:
fix build and do not forget on 10.0 this needs to be prompted to packagers

TayouVR commented:
can we have a small (at least equivalent to 1px of QR code) white border around the QR image? 
Could be possibly important with some themes and QR scanners. e.g. if your theme has a pure black background idk how well that'd work.

Also would be prettier.

9.x impl also has it:
![image](https://github.com/user-attachments/assets/0d8273f8-78db-4515-9b69-6f68d5e24546)

getchoo commented:
I think I got it to be about the same

<img width="372" height="514" alt="image" src="https://github.com/user-attachments/assets/267c39d1-7769-4aa2-a00f-8ea9b451599d" />

getchoo committed:
build(vcpkg): use a minimal libqrencode build

`libpng` in vcpkg doesn't have the right patches (libpng@ed68998) to
build with universal binary support on macOS, so let's avoid building it
(we need to bump our baseline for this too)

Thanks for letting me know it was optional, Kode :)

Signed-off-by: Seth Flynn <getchoo@tuta.io>

getchoo commented:
macOS builds strike again....I'll fix it later

TheKodeToad reviewed (changes requested):
I'm a little skeptical of all the introduced cmake stuff - shouldn't the goal be to reduce complexity and maintenance burden?
Would these patches be prone to breaking? At this point, what are we solving by avoiding vendoring for this tiny library?

TheKodeToad commented:
libpng doesn't seem to be a hard requirement:
> While the command-line utility and some test programs use libpng or SDL 2.0, the libqrencode library itself has no dependencies

getchoo commented:
> libpng doesn't seem to be a hard requirement:
> 
> > While the command-line utility and some test programs use libpng or SDL 2.0, the libqrencode library itself has no dependencies

This saved me

getchoo commented:
Rebased

<img width="372" height="514" alt="image" src="https://github.com/user-attachments/assets/1c387628-2f16-4732-953e-63ceb3963607" />

Trial97 commented on COPYING.md:
```diff
@@ -404,15 +404,6 @@
     You should have received a copy of the GNU Lesser General Public
     License along with this library. If not, see <http://www.gnu.org/licenses/>.
 
-## QR-Code-generator (`libraries/qrcodegenerator`)
```
Shouldn't this be replaced with the libqrencode copyright?

TheKodeToad commented on launcher/ui/dialogs/MSALoginDialog.cpp:
```diff
@@ -146,27 +146,32 @@ void MSALoginDialog::authorizeWithBrowser(const QUrl& url)
-    // NOTE: For performance reasons my implementation only draws the foreground parts in supplied color.
-    // It expects background to be prepared already (in white or whatever is preferred).
+    const auto* qr = QRcode_encodeString(data.toUtf8().constData(), 0, QRecLevel::QR_ECLEVEL_M, QRencodeMode::QR_MODE_8, 1);
+    if(!qr) {
```
Looks like somebody forget to format
another nitpick: I think it would be nicer to type some of these variables as auto doesn't really save much space (looks like somebody is a rust developer)

getchoo replied:
> Looks like somebody forget to format

Yes

> another nitpick: I think it would be nicer to type some of these variables as auto doesn't really save much space (looks like somebody is a rust developer)

naur

getchoo commented:
Hopefully the above will be merged quickly so we don't need any of this vcpkg stuff to hit develop.......

getchoo commented:
@TheKodeToad vcpkg stuff was thanos snapped

DioEgizio reviewed (approved):
just tried, works perfectly on Mac too!

TheKodeToad commented on launcher/ui/dialogs/MSALoginDialog.cpp:
```diff
@@ -146,27 +146,32 @@ void MSALoginDialog::authorizeWithBrowser(const QUrl& url)
-    // It expects background to be prepared already (in white or whatever is preferred).
+    const auto* qr = QRcode_encodeString(data.toUtf8().constData(), 0, QRecLevel::QR_ECLEVEL_M, QRencodeMode::QR_MODE_8, 1);
+    if (!qr) {
+        qWarning() << "Unable to encode '" << data << "' as QR code";
```
After intense scrutiny I found another issue - qt adds the spaces and quotes automatically
```suggestion
        qWarning() << "Unable to encode" << data << "as QR code";
```
(as if this code will ever run anyway)

TheKodeToad reviewed (approved):
Apart from that it works for me and my phone
